src/java/org/apache/cassandra/db/streaming/CassandraBlockStreamWriter.java

```diff
@@ -0,0 +1,109 @@
+ */
+public class CassandraBlockStreamWriter implements IStreamWriter
+{
+    private static final Logger logger = LoggerFactory.getLogger(CassandraBlockStreamWriter.class);
+
+    private final SSTableReader sstable;
+    private final ComponentManifest manifest;
+    private final StreamSession session;
+    private final StreamRateLimiter limiter;
+
+    public CassandraBlockStreamWriter(SSTableReader sstable, StreamSession session, ComponentManifest manifest)
+    {
+        this.session = session;
+        this.sstable = sstable;
+        this.manifest = manifest;
+        this.limiter =  StreamManager.getRateLimiter(session.peer);
+    }
+
+    /**
+     * Stream the entire file to given channel.
+     * <p>
+     *
+     * @param output where this writes data to
+     * @throws IOException on any I/O error
+     */
+    @Override
+    public void write(DataOutputStreamPlus output) throws IOException
+    {
+        long totalSize = manifest.getTotalSize();
+        logger.debug("[Stream #{}] Start streaming sstable {} to {}, repairedAt = {}, totalSize = {}", session.planId(),
+                     sstable.getFilename(), session.peer, sstable.getSSTableMetadata().repairedAt, totalSize);
+
+        long progress = 0L;
+        ByteBufDataOutputStreamPlus byteBufDataOutputStreamPlus = (ByteBufDataOutputStreamPlus) output;
+
+        for (Component component : manifest.getComponents())
+        {
+            @SuppressWarnings("resource") // this is closed after the file is transferred by ByteBufDataOutputStreamPlus
+            FileChannel in = new RandomAccessFile(sstable.descriptor.filenameFor(component), "r").getChannel();
+
+            // Total Length to transmit for this file
+            long length = in.size();
+
+            // tracks write progress
+            long bytesRead = 0;
+            logger.debug("[Stream #{}] Block streaming {}.{} gen {} component {} size {}", session.planId(),
+                        sstable.getKeyspaceName(), sstable.getColumnFamilyName(), sstable.descriptor.generation, component, length);
+
+            bytesRead += byteBufDataOutputStreamPlus.writeToChannel(in, limiter);
+            progress += bytesRead;
+
+            session.progress(sstable.descriptor.filenameFor(component), ProgressInfo.Direction.OUT, bytesRead,
+                             length);
+
+            logger.debug("[Stream #{}] Finished block streaming {}.{} gen {} component {} to {}, xfered = {}, length = {}, totalSize = {}",
+                         session.planId(), sstable.getKeyspaceName(), sstable.getColumnFamilyName(),
+                         sstable.descriptor.generation, component, session.peer, FBUtilities.prettyPrintMemory(bytesRead),
+                         FBUtilities.prettyPrintMemory(length), FBUtilities.prettyPrintMemory(totalSize));
+
+            byteBufDataOutputStreamPlus.flush();
+
+            logger.debug("[Stream #{}] Finished block streaming sstable {} to {}, xfered = {}, totalSize = {}",
```

Comment:
I believe this was supposed to go outside of the for-loop as well (:

Reply:
Done.